File: client/tests/roster.test.js
import { describe, expect, it } from "vitest";

import { ROSTER_STATES, dayPunches, daysOf, monthRoster, spanHours } from "../src/lib/roster.js";

/* ---------------------------------------------------------------------------
   One person's month.

   These rows are read to decide whether somebody was paid for a day, so every
   rule that decides one gets a test named after the rule. The order the day's
   status is settled in belongs to `rules.js` and is tested there; what is
   tested here is everything that sits around it — the times, the hours, the
   weekly-off half of "holiday", and the correction against the day.

   Relative import and no React: this runs the same functions the page does, in
   Node. CLAUDE.md §3.
   --------------------------------------------------------------------------- */

const IN = (t) => ({ log_type: "IN", time: t });
const OUT = (t) => ({ log_type: "OUT", time: t });

/** A month with nothing in it, so each test adds only what it is about. */
const base = {
	ym: "2026-08",
	punches: [],
	leave: [],
	holidays: [],
	corrections: [],
	shift: "Office Shift",
	today: "2026-09-01",
};

const dayOn = (out, iso) => out.rows.find((r) => r.iso === iso);

describe("daysOf", () => {
	it("gives every day of the month and stops at the last one", () => {
		expect(daysOf("2026-08")).toHaveLength(31);
		expect(daysOf("2026-09")).toHaveLength(30);
	});

	it("knows February in a leap year from February in an ordinary one", () => {
		expect(daysOf("2024-02")).toHaveLength(29);
		expect(daysOf("2026-02")).toHaveLength(28);
	});

	it("hands back nothing for a cycle that is not one", () => {
		expect(daysOf("")).toEqual([]);
		expect(daysOf("nonsense")).toEqual([]);
	});
});

describe("dayPunches", () => {
	it("takes the earliest in and the latest out, not the first two rows", () => {
		/* A gate that double-reads puts four rows on a day. Taken in arrival
		   order the second read is the punch-out, and a nine-hour day is
		   reported as four minutes. */
		const { inAt, outAt } = dayPunches([
			IN("2026-08-03 08:22:00"),
			IN("2026-08-03 08:26:00"),
			OUT("2026-08-03 17:25:00"),
			OUT("2026-08-03 17:29:00"),
		]);
		expect(inAt).toBe("2026-08-03 08:22:00");
		expect(outAt).toBe("2026-08-03 17:29:00");
	});

	it("treats a log type it does not recognise as an arrival", () => {
		/* The bridge writes IN and OUT; a device configured by somebody else can
		   write neither. Counting an unknown type as an arrival is the safe
		   rounding — it makes somebody present rather than absent. */
		expect(dayPunches([{ log_type: "", time: "2026-08-03 08:00:00" }]).inAt)
			.toBe("2026-08-03 08:00:00");
	});

	it("says nothing for a day with no punches", () => {
		expect(dayPunches([])).toEqual({ inAt: "", outAt: "" });
	});
});

describe("spanHours", () => {
	it("counts hours and minutes, not a clock time", () => {
		expect(spanHours("2026-08-03 08:22:00", "2026-08-03 17:25:00")).toBe("09:03");
	});

	it("says nothing when a punch is missing, rather than counting from zero", () => {
		expect(spanHours("", "2026-08-03 17:25:00")).toBe("");
		expect(spanHours("2026-08-03 08:22:00", "")).toBe("");
	});

	it("says nothing when the out is before the in", () => {
		/* A shift crossing midnight lands here, and a negative span drawn as
		   `-1:00` reads as a rule nobody wrote. Night shifts belong to the day
		   they started (CLAUDE.md §5) and this screen does not yet resolve them,
		   so it declines rather than guessing. */
		expect(spanHours("2026-08-03 22:00:00", "2026-08-03 06:00:00")).toBe("");
	});
});

describe("the month", () => {
	it("draws a row for every day, whether anything happened on it or not", () => {
		expect(monthRoster(base).rows).toHaveLength(31);
	});

	it("counts a day with both punches as a full day, with its hours", () => {
		const out = monthRoster({
			...base,
			punches: [IN("2026-08-03 08:22:00"), OUT("2026-08-03 17:25:00")],
		});
		const d = dayOn(out, "2026-08-03");
		expect(d.display).toBe("full");
		expect(d.hours).toBe("09:03");
		expect(out.counts.full).toBe(1);
	});

	it("counts a day with one punch as partial, not as a full day", () => {
		/* A day counts as worked only with a punch in *and* out. An open shift is
		   not a complete day and is not payroll-ready. */
		const out = monthRoster({ ...base, punches: [IN("2026-08-03 08:22:00")] });
		expect(dayOn(out, "2026-08-03").display).toBe("partial");
		expect(dayOn(out, "2026-08-03").hours).toBe("");
	});

	it("separates a weekly off from a named holiday, which the rule does not", () => {
		/* Both are "nobody's absence" to `resolveDayStatus`, and the same answer
		   to that question. They are different colours on the key and different
		   things to a person reading a month. */
		const out = monthRoster({
			...base,
			holidays: [
				{ holiday_date: "2026-08-02", description: "Sunday", weekly_off: 1 },
				{ holiday_date: "2026-08-15", description: "Independence Day", weekly_off: 0 },
			],
		});
		expect(dayOn(out, "2026-08-02").display).toBe("weekoff");
		expect(dayOn(out, "2026-08-15").display).toBe("holiday");
		expect(dayOn(out, "2026-08-15").holiday).toBe("Independence Day");
	});

	it("lets a punch beat an approved leave record", () => {
		/* Somebody who cancelled their leave and came in must not be marked as on
		   leave because the request was never withdrawn. The rule's first clause,
		   seen through this screen. */
		const out = monthRoster({
			...base,
			punches: [IN("2026-08-10 08:00:00"), OUT("2026-08-10 17:00:00")],
			leave: [{ from_date: "2026-08-10", to_date: "2026-08-10", status: "Approved", leave_type: "Casual Leave" }],
		});
		expect(dayOn(out, "2026-08-10").display).toBe("full");
	});

	it("keeps a request nobody has decided out of the leave column", () => {
		const out = monthRoster({
			...base,
			leave: [{ from_date: "2026-08-11", to_date: "2026-08-11", status: "Open", leave_type: "Sick Leave" }],
		});
		expect(dayOn(out, "2026-08-11").display).toBe("unappr");
	});

	it("covers every day a leave spans, including one that began the month before", () => {
		/* The read asks for overlapping applications rather than contained ones,
		   and this is what that is for: a leave from 30 July to 3 August covers
		   three days of this month. */
		const out = monthRoster({
			...base,
			leave: [{ from_date: "2026-07-30", to_date: "2026-08-03", status: "Approved", leave_type: "Earned Leave" }],
		});
		expect(dayOn(out, "2026-08-01").display).toBe("leave");
		expect(dayOn(out, "2026-08-03").display).toBe("leave");
		expect(dayOn(out, "2026-08-04").display).not.toBe("leave");
	});

	it("does not call a day absent before it has happened", () => {
		/* Treating a missing record as an absence would mark the whole group
		   absent at nine in the morning. */
		const out = monthRoster({ ...base, today: "2026-08-10" });
		expect(dayOn(out, "2026-08-09").display).toBe("absent");
		expect(dayOn(out, "2026-08-10").display).toBe("unmarked");
		expect(dayOn(out, "2026-08-20").display).toBe("unmarked");
	});

	it("hangs a correction on the day it is about", () => {
		const out = monthRoster({
			...base,
			corrections: [{
				name: "AR-1", attendance_date: "2026-08-19", status: "Pending Approval",
				requested_in: "2026-08-19 09:05:00", requested_out: "2026-08-19 18:10:00",
			}],
		});
		expect(dayOn(out, "2026-08-19").ar.name).toBe("AR-1");
		expect(dayOn(out, "2026-08-18").ar).toBeNull();
	});

	it("counts every day into exactly one state", () => {
		/* The strip along the top has to add up to the month, or it is a set of
		   numbers nobody can check. */
		const out = monthRoster({
			...base,
			today: "2026-09-01",
			punches: [IN("2026-08-03 08:00:00"), OUT("2026-08-03 17:00:00"), IN("2026-08-04 08:00:00")],
			holidays: [{ holiday_date: "2026-08-02", description: "Sunday", weekly_off: 1 }],
			leave: [{ from_date: "2026-08-05", to_date: "2026-08-05", status: "Approved" }],
		});
		const total = ROSTER_STATES.reduce((t, st) => t + (out.counts[st.key] || 0), 0);
		expect(total).toBe(31);
	});

	it("survives a month the site answered nothing for", () => {
		const out = monthRoster({ ...base, punches: null, leave: null, holidays: null, corrections: null });
		expect(out.rows).toHaveLength(31);
		expect(out.counts.absent).toBe(31);
	});
});

describe("a day carrying more than one leave application", () => {
	it("takes the granted one, whatever order the site listed them in", () => {
		/* Somebody applies, is refused, applies again and is granted. The site
		   lists the older one first, and taking the first would read a granted
		   day as absent — on a screen somebody is paid from. */
		const out = monthRoster({
			...base,
			leave: [
				{ from_date: "2026-08-12", to_date: "2026-08-12", status: "Rejected", leave_type: "Casual Leave" },
				{ from_date: "2026-08-12", to_date: "2026-08-12", status: "Approved", leave_type: "Sick Leave" },
			],
		});
		const d = dayOn(out, "2026-08-12");
		expect(d.display).toBe("leave");
		expect(d.leave.leave_type).toBe("Sick Leave");
	});

	it("prefers one still waiting over one already refused", () => {
		const out = monthRoster({
			...base,
			leave: [
				{ from_date: "2026-08-12", to_date: "2026-08-12", status: "Rejected" },
				{ from_date: "2026-08-12", to_date: "2026-08-12", status: "Open" },
			],
		});
		expect(dayOn(out, "2026-08-12").display).toBe("unappr");
	});

	it("carries a refusal alongside the day rather than dropping it", () => {
		/* "Asked, was refused, did not come in" is exactly the context HR wants
		   when they open the day — so the row keeps the application even though
		   the status is absent. */
		const out = monthRoster({
			...base,
			leave: [{ from_date: "2026-08-13", to_date: "2026-08-13", status: "Rejected", leave_type: "Casual Leave" }],
		});
		const d = dayOn(out, "2026-08-13");
		expect(d.display).toBe("absent");
		expect(d.leave.status).toBe("Rejected");
	});
});

File: app/test/roster_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:manna_hr_app/core/roster.dart';

/// The month, and the order every innocent explanation is checked in.
///
/// The same rules as `manna_hr/tests/test_rules.py` and the dashboard's
/// `roster.test.js`, restated here rather than assumed, because three screens
/// disagreeing about whether a punch beats an approved leave record would be
/// three answers to what somebody is paid.
void main() {
  Map<String, dynamic> punch(String time, String type) =>
      {'time': time, 'log_type': type};

  group('where somebody stood on a day', () {
    test('a punch beats an approved leave record', () {
      // Somebody who cancelled their leave and came in must not be marked
      // absent, or on leave, because the request was never withdrawn.
      expect(
        resolveDayStatus(
          hasPunchIn: true,
          hasPunchOut: true,
          leaveStatus: 'Approved',
          isHoliday: false,
          isPastDay: true,
        ),
        DayStatus.present,
      );
    });

    test('a punch in with no punch out is still in, not a full day', () {
      expect(
        resolveDayStatus(
          hasPunchIn: true,
          hasPunchOut: false,
          leaveStatus: '',
          isHoliday: false,
          isPastDay: true,
        ),
        DayStatus.onFloor,
      );
    });

    test('leave nobody has decided is not time off', () {
      expect(
        resolveDayStatus(
          hasPunchIn: false,
          hasPunchOut: false,
          leaveStatus: 'Open',
          isHoliday: false,
          isPastDay: true,
        ),
        DayStatus.leavePending,
      );
    });

    test('a refused leave request explains nothing, so the day is absent', () {
      expect(
        resolveDayStatus(
          hasPunchIn: false,
          hasPunchOut: false,
          leaveStatus: 'Rejected',
          isHoliday: false,
          isPastDay: true,
        ),
        DayStatus.absent,
      );
    });

    test('a holiday is nobody\'s absence', () {
      expect(
        resolveDayStatus(
          hasPunchIn: false,
          hasPunchOut: false,
          leaveStatus: '',
          isHoliday: true,
          isPastDay: true,
        ),
        DayStatus.holiday,
      );
    });

    test('a day that has not finished is unmarked, never absent', () {
      // Treating a missing record as an absence would mark the whole group
      // absent at nine in the morning.
      expect(
        resolveDayStatus(
          hasPunchIn: false,
          hasPunchOut: false,
          leaveStatus: '',
          isHoliday: false,
          isPastDay: false,
        ),
        DayStatus.unmarked,
      );
    });
  });

  group('which leave application covers a day', () {
    test('a granted application beats a refused one on the same day', () {
      // The site usually lists the older one first, and the older one is
      // usually the refused one. Taking it would read a granted day as absent.
      final lv = leaveOn('2026-09-09', [
        {'from_date': '2026-09-09', 'to_date': '2026-09-09', 'status': 'Rejected'},
        {'from_date': '2026-09-09', 'to_date': '2026-09-09', 'status': 'Approved'},
      ]);
      expect(lv?['status'], 'Approved');
    });

    test('a leave that started last month still covers days in this one', () {
      final lv = leaveOn('2026-09-02', [
        {'from_date': '2026-08-28', 'to_date': '2026-09-04', 'status': 'Approved'},
      ]);
      expect(lv, isNotNull);
    });
  });

  group('the month', () {
    test('every day of the month is a row, including the empty ones', () {
      expect(daysOfMonth('2026-09').length, 30);
      expect(daysOfMonth('2026-02').length, 28);
      expect(daysOfMonth('2024-02').length, 29);
    });

    test('a Sunday on the holiday list is a weekly off, not a holiday', () {
      // Two colours on the key, one answer to "was this person absent". The
      // split has to happen after the rule, not inside it.
      final rows = monthRoster(
        ym: '2026-09',
        punches: const [],
        leave: const [],
        holidays: [
          {'holiday_date': '2026-09-06', 'description': 'Sunday', 'weekly_off': 1},
          {'holiday_date': '2026-09-07', 'description': 'Onam', 'weekly_off': 0},
        ],
        corrections: const [],
        today: '2026-09-30',
      );
      expect(rows.firstWhere((r) => r.iso == '2026-09-06').status,
          DayStatus.weeklyOff);
      expect(rows.firstWhere((r) => r.iso == '2026-09-07').status,
          DayStatus.holiday);
    });

    test('a month read with no holiday list calls every Sunday absent', () {
      // Not a bug — the finding. It is why the calendar says out loud that no
      // holiday list is on the record, rather than drawing four red Sundays and
      // letting somebody conclude they were marked absent.
      final rows = monthRoster(
        ym: '2026-09',
        punches: const [],
        leave: const [],
        holidays: const [],
        corrections: const [],
        today: '2026-09-30',
      );
      expect(rows.firstWhere((r) => r.iso == '2026-09-06').status,
          DayStatus.absent);
    });

    test('the day carries its punches, its hours and its correction', () {
      final rows = monthRoster(
        ym: '2026-09',
        punches: [
          punch('2026-09-09 08:30:00', 'IN'),
          punch('2026-09-09 08:30:04', 'IN'),
          punch('2026-09-09 17:37:00', 'OUT'),
        ],
        leave: const [],
        holidays: const [],
        corrections: [
          {'attendance_date': '2026-09-10', 'status': 'Pending Approval'},
        ],
        today: '2026-09-30',
      );
      final worked = rows.firstWhere((r) => r.iso == '2026-09-09');
      expect(worked.status, DayStatus.present);
      expect(worked.hours, '09:07');
      expect(worked.punches.length, 3);
      expect(worked.correction, isNull);

      final asked = rows.firstWhere((r) => r.iso == '2026-09-10');
      expect(asked.correction?['status'], 'Pending Approval');
    });

    test('today is not absent, and the days after it are not either', () {
      final rows = monthRoster(
        ym: '2026-09',
        punches: const [],
        leave: const [],
        holidays: const [],
        corrections: const [],
        today: '2026-09-09',
      );
      expect(rows.firstWhere((r) => r.iso == '2026-09-09').status,
          DayStatus.unmarked);
      expect(rows.firstWhere((r) => r.iso == '2026-09-10').status,
          DayStatus.unmarked);
      expect(rows.firstWhere((r) => r.iso == '2026-09-08').status,
          DayStatus.absent);
    });

    test('every state on the key is counted, including the ones with none', () {
      final counts = countStates(monthRoster(
        ym: '2026-09',
        punches: const [],
        leave: const [],
        holidays: const [],
        corrections: const [],
        today: '2026-09-09',
      ));
      expect(counts.keys.length, DayStatus.values.length);
      expect(counts[DayStatus.onLeave], 0);
    });
  });
}
